logging/kit/options.go

```diff
@@ -0,0 +1,127 @@
+}
+
+// WithDecider customizes the function for deciding if the gRPC interceptor logs should log.
+func WithDecider(f grpc_logging.Decider) Option {
+	return func(o *options) {
+		o.shouldLog = f
+	}
+}
+
+// WithLevels customizes the function for mapping gRPC return codes and interceptor log level statements.
+func WithLevels(f CodeToLevel) Option {
+	return func(o *options) {
+		o.levelFunc = f
+	}
+}
+
+// WithCodes customizes the function for mapping errors to error codes.
+func WithCodes(f grpc_logging.ErrorToCode) Option {
+	return func(o *options) {
+		o.codeFunc = f
+	}
+}
+
+// WithDurationField customizes the function for mapping request durations to log fields.
+func WithDurationField(f DurationToField) Option {
+	return func(o *options) {
+		o.durationFunc = f
+	}
+}
+
+// DefaultCodeToLevel is the default implementation of gRPC return codes and interceptor log level for server side.
+func DefaultCodeToLevel(code codes.Code, logger log.Logger) log.Logger {
+	switch code {
+	case codes.OK, codes.Canceled, codes.InvalidArgument, codes.NotFound, codes.AlreadyExists, codes.Unauthenticated:
+		return level.Info(logger)
+	case codes.DeadlineExceeded, codes.PermissionDenied, codes.ResourceExhausted, codes.FailedPrecondition, codes.Aborted, codes.OutOfRange, codes.Unavailable:
+		return level.Warn(logger)
+	case codes.Unknown, codes.Unimplemented, codes.Internal, codes.DataLoss:
+		return level.Error(logger)
+	default:
+		return level.Error(logger)
+	}
+}
+
+// DefaultClientCodeToLevel is the default implementation of gRPC return codes to log levels for client side.
+func DefaultClientCodeToLevel(code codes.Code, logger log.Logger) log.Logger {
+	switch code {
+	case codes.OK, codes.Canceled, codes.InvalidArgument, codes.NotFound, codes.AlreadyExists, codes.ResourceExhausted, codes.FailedPrecondition, codes.Aborted, codes.OutOfRange:
+		return level.Debug(logger)
+	case codes.Unknown, codes.DeadlineExceeded, codes.PermissionDenied, codes.Unauthenticated:
+		return level.Info(logger)
+	case codes.Unimplemented, codes.Internal, codes.Unavailable, codes.DataLoss:
+		return level.Warn(logger)
+	default:
+		return level.Info(logger)
+	}
+}
+
+// DefaultDurationToField is the default implementation of converting request duration to a kit field.
+var DefaultDurationToField = DurationToTimeMillisField
+
+// DurationToTimeMillisField converts the duration to milliseconds and uses the key `grpc.time_ms`.
+func DurationToTimeMillisField(duration time.Duration) []interface{} {
+	return []interface{}{"grpc.time_ms", durationToMilliseconds(duration)}
+}
+
+// DurationToDurationField uses a Duration field to log the request duration
+// and leaves it up to Log's encoder settings to determine how that is output.
+func DurationToDurationField(duration time.Duration) []interface{} {
+	return []interface{}{"grpc.duration", duration}
+}
+
+func durationToMilliseconds(duration time.Duration) float32 {
+	return float32(duration.Nanoseconds()/1000) / 1000
```

Comment:
Why not `duration.Milliseconds`? https://golang.org/pkg/time/#Duration.Milliseconds

Reply:
The project is still built with Go 1.12, should we upgrade ?